### ISSUE TITLE
S3 Media Source for storing gallery images

### ISSUE THREAD

**Issue by mtwalsh** [splittingred/Gallery#256]
It would be great to be able to use the Gallery snippet with an S3 media source so images are held on Amazon S3 rather than on the server.  

Apologies if this is already possible (I couldn't spot where).

**Comment by oksushi** [comment]
+1 for this.

Perhaps some part of phpThumbOf could be used for this?

**Comment by mtwalsh** [comment]
@oksushi Thanks, yes you can serve the images from S3/Cloudfront using phpThumbOf but I need to store the images on S3 as we want to use Gallery on sites that are load-balanced so we can't have anything stored on the server itself other than the source code.

**Comment by oksushi** [comment]
Agreed — I was referring to reusing code from phpThumbOf, not for using it directly to resize images. Having source images and Gallery-resized images both stored on S3 would be helpful.

**Comment by mtwalsh** [comment]
@oksushi Ah, apologies.  Yes, although I don't think it will even need to interface with S3 itself, if it's just changed to work with an S3 media source hopefully the media source would then handle all of the heavy lifting in terms of get/put onto S3 as that's already built-in.

**Comment by C8NO28** [comment]
+1 for this.

**Comment by UzverNumber47** [comment]
+1

**Comment by lithiumlab** [comment]
+1

**Comment by aksteven** [comment]
Would definitely like to be able to use Gallery with CloudFront/S3. +1

**Comment by markwillis82** [comment]
I have just added a pull request with support for media sources for the gallery plugin. This allows S3/cloud files / etc to be used as the storage for images.

We needed to develop this feature to allow us to load balance our modx sites across multiple servers (and not using dirty hacks to rsync images between them)

PR: https://github.com/splittingred/Gallery/pull/269

**Comment by mtwalsh** [comment]
Great news, thanks Mark.

**Comment by mtwalsh** [comment]
Does anyone know what the process is to get this pull request integrated and the package updated?  Would be good to get it done ASAP.
